File: benchmarks/ruby/quicksort.rb
def swap(a, i, j)
  a[i], a[j] = a[j], a[i]
end

def quicksort(a, lo, hi)
  return if lo >= hi

  pivot = a[(lo + hi) / 2]
  i = lo
  j = hi

  while i <= j
    i += 1 while a[i] < pivot
    j -= 1 while a[j] > pivot
    if i <= j
      swap(a, i, j)
      i += 1
      j -= 1
    end
  end

  quicksort(a, lo, j)
  quicksort(a, i, hi)
end

n = 10_000
data = Array.new(n) { |i| n - i }

quicksort(data, 0, n - 1)
puts data[5000]

File: benchmarks/java/Quicksort.java
public class Quicksort {
    static void swap(int[] a, int i, int j) {
        int t = a[i];
        a[i] = a[j];
        a[j] = t;
    }

    static void quicksort(int[] a, int lo, int hi) {
        if (lo >= hi) return;

        int pivot = a[(lo + hi) >>> 1];
        int i = lo;
        int j = hi;

        while (i <= j) {
            while (a[i] < pivot) i++;
            while (a[j] > pivot) j--;
            if (i <= j) {
                swap(a, i, j);
                i++;
                j--;
            }
        }

        quicksort(a, lo, j);
        quicksort(a, i, hi);
    }

    public static void main(String[] args) {
        int n = 10000;
        int[] data = new int[n];
        for (int i = 0; i < n; i++) {
            data[i] = n - i;
        }

        quicksort(data, 0, n - 1);
        System.out.println(data[5000]);
    }
}
